File: Task4A_B/task4a_b/src/main/java/com/cps2004/Book.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class Book extends Product{
    private String publisher;
    private int pages;
    private String author;

    // Default constructor
    public Book(){
        super();
        publisher = "";
        pages = 0;
        author = "";
    }

    // Parameterized constructor
    public Book(String name, int id, Supplier supplier, int price, int volume, String publisher, int pages, String author){
        super(name, id, supplier, price, volume);
        this.publisher = publisher;
        this.pages = pages;
        this.author = author;
    }

    // Getters and setters
    public String getPublisher() { return publisher; }
    public int getPages() { return pages; }
    public String getAuthor() { return author; }
    public void setPublisher(String publisher) { this.publisher = publisher; }
    public void setPages(int pages) { this.pages = pages; }
    public void setAuthor(String author) { this.author = author; }

    @Override
    // Calculate discounted price
    public abstract double calculate_discount(String month, int quantity);

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product publisher: " + publisher);
        System.out.println("Product pages: " + pages);
        System.out.println("Product author: " + author);
    }
}

File: Task3C/book.cpp
#include "book.h"

// Default constructor
Book::Book() { 
    this->author = "";
    this->publisher = "";
    this->pages = 0;
}

// Parameterised constructor
Book::Book(string name, int price, Supplier* supplier, int volume, int productID, string author, string publisher, int pages) : Product(name, price, supplier, volume, productID) {
    this->author = author;
    this->publisher = publisher;
    this->pages = pages;
}

// Getters and setters
string Book::getAuthor() {
    return this->author;
}

string Book::getPublisher() {
    return this->publisher;
}

int Book::getPages() {
    return this->pages;
}

void Book::setAuthor(string author) {
    this->author = author;
}

void Book::setPublisher(string publisher) {
    this->publisher = publisher;
}

void Book::setPages(int pages) {
    this->pages = pages;
}

// Display product information
void Book::display_product_info() {
    Product::display_product_info();
    cout << "Author: " << this->author << endl;
    cout << "Publisher: " << this->publisher << endl;
    cout << "Pages: " << this->pages << endl;
}

// Save product information to file
void Book::save(ostream& out) const {
    Product::save(out);
    out << publisher << "," << pages << "," << author << "\n";
}

// Load product information from file
void Book::load(istream& in) {
    Product::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    publisher = token;

    getline(ss, token, ',');
    pages = stoi(token);

    getline(ss, token, ',');
    author = token;
}